backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/bpartner/relation/BPRelationsService.java

```diff
@@ -0,0 +1,229 @@
+		final IdentifierString targetBpIdentifier = IdentifierString.of(relatedBp.getTargetBpartnerIdentifier());
+		final BPartnerComposite targetBpartnerComposite =
+				Check.assumePresent(jsonServiceFactory.createRetriever().getBPartnerComposite(sourceBPartner.getOrgId(), targetBpIdentifier)
+						, "No bpartner found for identifier: " + targetBpIdentifier);
+
+		final BPartnerId targetBPartnerId = targetBpartnerComposite.getBpartner().getId();
+		final IdentifierString targetLocationIdentifier = IdentifierString.ofOrNull(relatedBp.getTargetLocationIdentifier());
+		final Optional<BPartnerLocation> targetLocationOpt = getBpartnerLocation(targetBpIdentifier, targetLocationIdentifier, targetBpartnerComposite);
+		final BPartnerLocation targetLocation = Check.assumeNotNull(targetLocationOpt.orElseGet(getTargetPartnerLocationForRelation(targetBpartnerComposite, relatedBp.getShipTo(), relatedBp.getBillTo())), "Cannot identify location for target BPartner: {}, LocationIdentifier: {}", targetBpIdentifier, targetLocationIdentifier);
+
+		final BPRelation.BPRelationBuilder builder = BPRelation.builder()
+				.bpartnerId(bpartnerId)
+				.targetBPartnerId(targetBPartnerId)
+				.targetBPLocationId(targetLocation.getId())
+				.name(relatedBp.getName())
+				.description(relatedBp.getDescription())
+				.billTo(relatedBp.getBillTo())
+				.fetchedFrom(relatedBp.getFetchedFrom())
+				.handoverLocation(relatedBp.getHandOverLocation())
+				.payFrom(relatedBp.getPayFrom())
+				.remitTo(relatedBp.getRemitTo())
+				.shipTo(relatedBp.getShipTo())
+				.active(relatedBp.getActive());
+		if (relatedBp.getRole() != null)
+		{
+			builder.role(BPRelation.BPRelationRole.valueOf(relatedBp.getRole().name()));
+		}
+		if (relatedBp.getExternalId() != null)
+		{
+			builder.externalId(ExternalId.ofOrNull(relatedBp.getExternalId()));
+		}
+		location.ifPresent(loc -> builder.bpLocationId(loc.getId()));
+		return builder.build();
+	}
+
+	private Supplier<? extends BPartnerLocation> getTargetPartnerLocationForRelation(@NonNull final BPartnerComposite bpartnerComposite, @Nullable final Boolean isShipTo, @Nullable final Boolean isBillTo)
+	{
+		Optional<BPartnerLocation> location = Optional.empty();
+		if (Boolean.TRUE.equals(isBillTo))
+		{
+			location = bpartnerComposite.extractBillToLocation();
+			if (!location.isPresent())
+			{
+				location = bpartnerComposite.extractLocation(l -> l.getLocationType().getIsBillToOr(false));
+			}
+		}
+		if (Boolean.TRUE.equals(isShipTo))
+		{
+			location = bpartnerComposite.extractShipToLocation();
+			if (!location.isPresent())
+			{
+				location = bpartnerComposite.extractLocation(l -> l.getLocationType().getIsShipToOr(false));
+			}
+		}
+		if (!location.isPresent())
+		{
+			location = bpartnerComposite.extractLocation(Objects::nonNull);
+		}
+		final Optional<BPartnerLocation> result = location;
+		return () -> Check.assumePresent(result, "Cannot infer location for source BPartner");
```

Comment:
please throw a MissingPropertyException instead (rationale: it turned out that request is too unspecific as it lacks the explicit targetLocation which is needed for this particular bpartner)

Reply:
Makes sense. I also noticed that the wording still referenced the source partner location, while the logic is applied only to the target bpartner. Updated to throw MissingPropertyException  and updated wording.